File: clients/python/highbar_client/session.py
# SPDX-License-Identifier: GPL-2.0-only
"""Hello handshake + AI token loader (T086).

Mirrors clients/fsharp/src/Session.fs behavior so SC-004's byte-equal
stream claim holds: both clients ship the same SchemaVersion, both
do strict-equality verification on HelloResponse, both read the token
file with exponential backoff up to 5s.
"""

from __future__ import annotations

import enum
import os
import time
from dataclasses import dataclass
from typing import Optional

import grpc

from . import SCHEMA_VERSION
from .highbar import service_pb2, service_pb2_grpc, state_pb2  # type: ignore

TOKEN_HEADER = "x-highbar-ai-token"


class ClientRole(enum.Enum):
    OBSERVER = service_pb2.Role.ROLE_OBSERVER
    AI = service_pb2.Role.ROLE_AI


@dataclass(slots=True)
class Handshake:
    session_id: str
    schema_version: str
    static_map: state_pb2.StaticMap
    current_frame: int
    role: ClientRole


def read_token_with_backoff(path: str, max_delay_ms: int = 5000) -> str:
    """Read the AI token file, retrying with exponential backoff.

    The plugin writes the token after init but before Bind unblocks —
    a client that starts simultaneously can race it. Matches the F#
    client's 25ms → 1000ms cap; total wait bounded by ``max_delay_ms``.
    """
    deadline = time.monotonic() + (max_delay_ms / 1000.0)
    delay_ms = 25
    while not os.path.isfile(path):
        if time.monotonic() >= deadline:
            raise TimeoutError(
                f"token file not present within {max_delay_ms}ms: {path}"
            )
        time.sleep(delay_ms / 1000.0)
        delay_ms = min(1000, delay_ms * 2)
    with open(path, "r", encoding="utf-8") as f:
        return f.read().strip()


def hello(
    channel: grpc.Channel,
    role: ClientRole,
    client_id: str,
    token: Optional[str] = None,
) -> Handshake:
    """Open a Hello RPC. Raises grpc.RpcError on FAILED_PRECONDITION /
    PERMISSION_DENIED / ALREADY_EXISTS per contracts/README.md §Hello.

    ``token``: required when ``role == AI``. The server's auth
    interceptor does not gate Hello on the token, but a token-carrying
    Hello keeps the metadata handle available for subsequent RPCs on
    the same channel.
    """
    stub = service_pb2_grpc.HighBarProxyStub(channel)
    req = service_pb2.HelloRequest(
        schema_version=SCHEMA_VERSION,
        client_id=client_id,
        role=role.value,
    )
    metadata: list[tuple[str, str]] = []
    if token is not None:
        metadata.append((TOKEN_HEADER, token))

    resp = stub.Hello(req, metadata=metadata)

    # FR-022a defense-in-depth mirror of the F# client: server already
    # rejected mismatch with FAILED_PRECONDITION, but re-verify here so
    # unit tests that stub the server can't silently drift.
    if resp.schema_version != SCHEMA_VERSION:
        raise RuntimeError(
            f"schema mismatch: server={resp.schema_version} "
            f"client={SCHEMA_VERSION}"
        )

    return Handshake(
        session_id=resp.session_id,
        schema_version=resp.schema_version,
        static_map=resp.static_map,
        current_frame=resp.current_frame,
        role=role,
    )


def get_command_schema(
    channel: grpc.Channel,
    token: str,
    timeout: Optional[float] = None,
) -> service_pb2.CommandSchemaResponse:
    stub = service_pb2_grpc.HighBarProxyStub(channel)
    return stub.GetCommandSchema(
        service_pb2.CommandSchemaRequest(),
        metadata=[(TOKEN_HEADER, token)],
        timeout=timeout,
    )


def get_unit_capabilities(
    channel: grpc.Channel,
    token: str,
    unit_id: int,
    timeout: Optional[float] = None,
) -> service_pb2.UnitCapabilitiesResponse:
    stub = service_pb2_grpc.HighBarProxyStub(channel)
    req = service_pb2.UnitCapabilitiesRequest(unit_id=unit_id)
    return stub.GetUnitCapabilities(
        req,
        metadata=[(TOKEN_HEADER, token)],
        timeout=timeout,
    )

File: clients/fsharp/src/Session.fs
// SPDX-License-Identifier: GPL-2.0-only
//
// HighBarV3 — Hello handshake (T043, T045).
//
// Thin wrapper around the generated HighBarProxy client that
// performs the FR-022a strict-equality schema version check and
// surfaces the resulting session_id + static_map to the caller. AI
// role (token) support lands in US2 T063; this module handles the
// observer-role and token-authenticated paths uniformly.

namespace HighBar.Client

open System
open System.IO
open System.Threading
open System.Threading.Tasks
open Grpc.Core
open Grpc.Net.Client
open Highbar.V1

module Session =

    /// Schema version baked into the client. MUST match the
    /// compile-time constant on the plugin side
    /// (src/circuit/grpc/SchemaVersion.h). Any mismatch is a
    /// deployment error — loud, fast failure is the point.
    [<Literal>]
    let SchemaVersion = "1.0.0"

    type ClientRole = Observer | Ai

    /// Attempt to read the AI token file with exponential backoff
    /// (handles the plugin-startup race per spec Edge Case / data-
    /// model §7). Throws on timeout.
    let readTokenWithBackoff (path: string) (maxDelayMs: int) : Async<string> =
        async {
            let deadline = DateTime.UtcNow.AddMilliseconds(float maxDelayMs)
            let mutable delayMs = 25
            while not (File.Exists(path)) do
                if DateTime.UtcNow >= deadline then
                    failwithf "token file not present within %dms: %s" maxDelayMs path
                do! Async.Sleep(delayMs)
                delayMs <- min 1000 (delayMs * 2)
            return (File.ReadAllText(path).Trim())
        }

    /// Result of a successful handshake.
    type Handshake = {
        SessionId: string
        StaticMap: StaticMap
        CurrentFrame: uint32
        /// The ClientRole the handshake was opened with. Tokenless
        /// observers may still keep this to drive client-side RPC gating.
        Role: ClientRole
    }

    /// Open a Hello RPC. Returns Handshake on OK. FAILED_PRECONDITION
    /// from the server bubbles as an RpcException — callers should
    /// surface the status detail (contains both server and client
    /// schema versions) directly to the user.
    let hello (channel: GrpcChannel)
              (role: ClientRole)
              (clientId: string)
              (tokenOpt: string option)
              : Async<Handshake> =
        async {
            let client = HighBarProxy.HighBarProxyClient(channel)
            let req = HelloRequest(
                        SchemaVersion = SchemaVersion,
                        ClientId = clientId,
                        Role = (match role with
                                | Observer -> Highbar.V1.Role.Observer
                                | Ai -> Highbar.V1.Role.Ai))
            let metadata = Metadata()
            match tokenOpt with
            | Some t -> metadata.Add("x-highbar-ai-token", t)
            | None -> ()

            let call = client.HelloAsync(req, metadata)
            let! resp = call.ResponseAsync |> Async.AwaitTask

            // FR-022a defense in depth: the server already rejected a
            // mismatch with FAILED_PRECONDITION, but re-verify here so
            // tests that mock the server can't silently drift.
            if resp.SchemaVersion <> SchemaVersion then
                failwithf "schema mismatch: server=%s client=%s"
                          resp.SchemaVersion SchemaVersion

            return {
                SessionId = resp.SessionId
                StaticMap = resp.StaticMap
                CurrentFrame = resp.CurrentFrame
                Role = role
            }
        }

    let getCommandSchema (channel: GrpcChannel)
                         (metadata: Metadata)
                         (ct: CancellationToken)
                         : Async<CommandSchemaResponse> =
        async {
            let client = HighBarProxy.HighBarProxyClient(channel)
            let call = client.GetCommandSchemaAsync(
                           CommandSchemaRequest(),
                           metadata,
                           cancellationToken = ct)
            return! call.ResponseAsync |> Async.AwaitTask
        }

    let getUnitCapabilities (channel: GrpcChannel)
                            (metadata: Metadata)
                            (unitId: uint32)
                            (ct: CancellationToken)
                            : Async<UnitCapabilitiesResponse> =
        async {
            let client = HighBarProxy.HighBarProxyClient(channel)
            let req = UnitCapabilitiesRequest(UnitId = unitId)
            let call = client.GetUnitCapabilitiesAsync(
                           req,
                           metadata,
                           cancellationToken = ct)
            return! call.ResponseAsync |> Async.AwaitTask
        }
